generator/wrapper_gen.py

```diff
@@ -175,11 +175,11 @@ def my_procid():
 
 def _get_data(x):
     if isinstance(x, pdDataFrame):
-        x_dtypes = x.dtypes
+        x_dtypes = x.dtypes.values
         if np.all(x_dtypes == x_dtypes[0]):
             x = x.to_numpy()
         else:
-            x = [xi.to_numpy() for _, xi in x.items()]
+            x = [np.asfortranarray(xi.values) for _, xi in x.items()]
```

Comment:
Back to values? Fortran?

Reply:
We have difficulties with this case. if you take on separate columns the data may be non-contiguous. So far the only way that has turned out is to use the function `np.asfortranarray`. I’m still looking at what can be done so as not to degrade and handle the whole case correctly. If there are ideas and suggestions - I will be very grateful.

Reply:
What about np.ascontiguousarray(xi.to_numpy)`?

Reply:
With `to_numpy` we have the result is slower a bit in this case ~2-3 ms.

Reply:
The user might be be able to avoid/change the non-homogen data-layout without doing a copy. Implicitly copying the DF with then make it impossible to process such data. 
If we support fragmented arrays without copies it might be slow but we would not lose support for larger datasets. The user can always decide to create a contiguous array if there is enough memory for small cases.

In other words, I suggest always preferring in-place operation if possible. Not making copies is the only reason why we actually provide support for DFs in addition to numpy arrays.

Reply:
I wrote some example that covers all interesting cases with a data frame. Ok, for simple case we have numpy array and we created data frame from array.

## 1 case. homogen & сontiguous data 
```python
fptype = np.float32
row = 10**5
col = 10**3
na1 = np.random.randn(row, col)
X_df = pd.DataFrame(na1, dtype=fptype)
```
Let's see what it is
```python
ordata = X_df._data.blocks[0].get_values()

print("1. Create DF. Data n blocks {}. Block[0]:  OWNDATA: {:2}; C_CONT: {:2}; F_CONT: {:2}; strides ({} {})".
    format(len(X_df._data.blocks), ordata.flags['OWNDATA'], ordata.flags['C_CONTIGUOUS'],
            ordata.flags['F_CONTIGUOUS'], ordata.strides[0],
            ordata.strides[1]))
```
Output:
```
1. Create DF. Data n blocks 1. Block[0]:  OWNDATA:  0; C_CONT:  0; F_CONT:  1; strides (4 4000)
```

1 case inside does not owner the data itself, uses the interface through the view. The data inside is displayed in fortran format. In this case, if we get values for DataFrame. We will get a view of this data.

```python
vdf = X_df.values
print("1.1. Get values. OWNDATA: {:2}; C_CONT: {:2}; F_CONT: {:2}; strides ({} {})"
    .format(vdf.flags['OWNDATA'], vdf.flags['C_CONTIGUOUS'],
            vdf.flags['F_CONTIGUOUS'], vdf.strides[0],
            vdf.strides[1]))
```
Output:
```
1.1. Get values.  OWNDATA:  0; C_CONT:  1; F_CONT:  0; strides (4000 4)
```
There is no copy. On the daal4pu side, will call `_make_hnt` for creating `HomogenTable`. This is the fastest and most free way to work with data! If it is created from a fortran numpy format, it will be created `SOATable`. 

## 2 case. homogen & non-contiguous data 

```python
X_df = X_df.iloc[2 : row - 1, 2 : col - 2]
```
What is going on inside:
```python
ordata = X_df._data.blocks[0].get_values()

print("2. Get non-contiguous DF. Data n blocks {}. Block[0]: OWNDATA: {:2}; C_CONT: {:2}; F_CONT: {:2}; strides ({} {})".
    format(len(X_df._data.blocks), ordata.flags['OWNDATA'], ordata.flags['C_CONTIGUOUS'],
            ordata.flags['F_CONTIGUOUS'], ordata.strides[0],
            ordata.strides[1]))
```
Output:
```
2. Get non-contiguous DF. Data n blocks 1. Block[0]: OWNDATA:  0; C_CONT:  0; F_CONT:  0; strides (4 4000)
```
We still do not own the data. It can be seen that the data is now not contiguous. What happens if in this case we take the data?

```python
vdf = X_df.values
print("2.1. Get values. OWNDATA: {:2}; C_CONT: {:2}; F_CONT: {:2}; strides ({} {})"
    .format(vdf.flags['OWNDATA'], vdf.flags['C_CONTIGUOUS'],
            vdf.flags['F_CONTIGUOUS'], vdf.strides[0],
            vdf.strides[1]))
```
Output:
```
2.1. Get values.  OWNDATA:  0; C_CONT:  0; F_CONT:  0; strides (4000 4)
```
Data in this case will not be owner. Those. In this case, **copying does not occur**. Moreover, data not-contiguous is preserved. So in this case it will be created `NpyNumericTable<NpyNonContigHandler>`. Which will lead to the correct result.  **Please note that now this case does not work correctly!**

## 3 case. heterogeneous & contiguous data 

```python
X_df = X_df.astype({X_df.columns[1]: 'float64', X_df.columns[6]: 'float64'})
X_df[X_df.columns[1]] = X_df[X_df.columns[5]] * 3
```

```python
ordata = X_df._data.blocks[0].get_values()
print("3. Make geterogenious DF. Data n blocks {}. Block[0]: OWNDATA: {:2}; C_CONT: {:2}; F_CONT: {:2}; strides ({} {})". format(len(X_df._data.blocks), ordata.flags['OWNDATA'], ordata.flags['C_CONTIGUOUS'],
            ordata.flags['F_CONTIGUOUS'], ordata.strides[0],
            ordata.strides[1]))
```
Output:
```
3. Make geterogenious DF. Data n blocks 3. Block[0]: OWNDATA:  1; C_CONT:  1; F_CONT:  0; strides (399988 4)
```
Somewhere inside there was a copy and now the data frame owns the data. With this data contiguous we have got. In this case, the valid list creation code `[xi.values for _, xi in x.items()]`. You can leave it. But there is potential to improve this. In daal4pu we created `SOA Table`

## 4 case. heterogeneous & not-contiguous data 

I could not reproduce this case. But in theory, it should work as case 3, only after converting it to fortran format. But I'm not sure if this case is possible at all?

## Summary

> The user might not be able to avoid/change the non-homogen data-layout without doing a copy. Implicitly copying the DF will then make it impossible to process such data.

In no case will data be copied. The pandas provides us is always a view of data. 1 case is the most popular in our cases and benchmarks. therefore, we would not want to have an overhead in this place. I agree that the rest cases are an important. Some do not work now. My current changes should solve these problems.

Reply:
Have you verified with looking at addresses if case 3 is really true? With the data you show I cannot verify the claim that there is no copy made.
I might just missing your point since I think, `X_df._data.blocks[0].get_values()`  should be `X_df[X_df.columns[1]]`.

Reply:
@fschlimb Yes, I checked.

Yes, I checked. When the data is homogeneous, they are stored in a single manager block. `X_df._data.blocks[0].get_values()` This is just an internal function of direct access to the memory of this block. `len(X_df._data.blocks)` - the number of such blocks. When we do a column addition or type change, new blocks will added. But at certain points, they are merged by type.
```python
X_df._data.blocks[0].get_values().data
```
Output:
```
<memory at 0x7f47c61928b8>
```
For get values:
```python
X_df.values.data
```
Output:
```
<memory at 0x7f47c61928b8>
```
Moreover, if copying would occur, we would get an contiguous array as a result. In this case, we get **not-contiguous**.